### PR TITLE
8054022: HttpURLConnection timeouts with Expect: 100-Continue and no chunking

### PR DESCRIPTION
**Pull request by DarraghClarke**
Currently it is possible for `HttpURLConnection` with the `Expect: 100-Continue` header to timeout awaiting for a server response. According to [RFC-7231](https://www.rfc-editor.org/rfc/rfc7231#section-5.1.1) a client `SHOULD NOT wait for an indefinite period before sending the message body`.

This PR changes the existing `expect100Continue` method to wait for a maximum of 5 seconds for a server response, this will be shorter if a timeout is set. If no response is received, the message is sent regardless.

Tests have been added to account for different scenarios that currently timeout, and the changes have been tested against tiers 1,2 and 3.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8054022](https://bugs.openjdk.org/browse/JDK-8054022): HttpURLConnection timeouts with Expect: 100-Continue and no chunking

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13330/head:pull/13330` \
`$ git checkout pull/13330`

Update a local copy of the PR: \
`$ git checkout pull/13330` \
`$ git pull https://git.openjdk.org/jdk.git pull/13330/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13330`

View PR using the GUI difftool: \
`$ git pr show -t 13330`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13330.diff">https://git.openjdk.org/jdk/pull/13330.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13330#issuecomment-1496325183)

### CODE REVIEW

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        control.response = "HTTP/1.1 200 OK\r\n" +
+                "Connection: close\r\n" +
+                "Content-Length: " + body.length() + "\r\n" +
+                "\r\n" +
+                body + "\r\n";
+
+        control.respondWith100Continue = true;
+        control.write100ContinueTwice = false;
+
+        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
```
Comment (justin-curtis-lu) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:193]: Hi Darragh,

It looks like the tests share a lot of similar code, unless there is a reason not to, could they not share a similar function to reduce code duplication?

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java
Comment (djelinski) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:1]: I noticed that you have tests for chunked streaming mode and for buffered mode; can you add tests for fixed length streaming mode as well? I.e. use `setFixedLengthStreamingMode`.

**Review by djelinski — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/sun/net/www/protocol/http/HttpURLConnection.java [hunk 2]
```diff
@@ -1350,14 +1351,23 @@ private void expect100Continue() throws IOException {
...
+                    // responseCode will be returned to caller
                     throw new ProtocolException("Server rejected operation");
                 }
             }
 
+            if (logger.isLoggable(PlatformLogger.Level.FINE)) {
+                logger.fine("Restoring original timeout : " + oldTimeout);
+            }
+
+            http.setIgnoreContinue(true);
```
[the remark this answers is not in the document]
Comment (DarraghClarke) [reply src/java.base/share/classes/sun/net/www/protocol/http/HttpURLConnection.java:1370]: Thanks for pointing those out, I accidentally doubled up on `setIgnoreContinue` calls

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java (hunk 1 above)
Comment (DarraghClarke) [reply test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:193]: Hey, thanks for pointing that out, I'll try and reduce some of the duplication

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+    class Control {
+        ServerSocket serverSocket = null;
+        boolean stop = false;
+        boolean respondWith100Continue = false;
+        boolean write100ContinueTwice = false;
+        String response = null;
+    }
+
+    private Thread serverThread = null;
+    private volatile Control control = null;
```
Comment (Michael-Mc-Mahon) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:61]: Afaik, the volatile reference to the Control object does not imply that references to its member fields are volatile, and they are being accessed from different threads.

**Comment by DarraghClarke** [comment]
Sorry for the delay in responding to the feedback, I was sidetracked on some other tasks then had an intermittent failure when I was adding the new fixed Length tests here.

I've added 3 tests to cover fixed length, tried to cut down some duplication by creating a shared `createConnection` method and cut down some of the server code.

Let me know if there are any other changes that should be made.

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    class Control {
+        volatile ServerSocket serverSocket = null;
+        volatile boolean stop = false;
+        volatile boolean respondWith100Continue = false;
+        volatile boolean write100ContinueTwice = false;
+        volatile String response = null;
+    }
+
+    private Thread serverThread = null;
+    private Control control = null;
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:61]:
```suggestion
    private volatile Control control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+
+    static {
+        logger = Logger.getLogger("sun.net.www.protocol.http.HttpURLConnection");
+        logger.setLevel(Level.ALL);
+        Logger.getLogger("").getHandlers()[0].setLevel(Level.ALL);
+    }
+
+    @BeforeAll
+    public void startServerSocket() throws Exception {
+        control = new Control();
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:72]:
```suggestion
        Control control = this.control = new Control();
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+        };
+        serverThread = new Thread(runnable);
+        serverThread.start();
+    }
+
+    @AfterAll
+    public void stopServerSocket() throws Exception {
+        control.stop = true;
+        control.serverSocket.close();
+        serverThread.join();
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:177]:
```suggestion
        Control control = this.control;
        control.stop = true;
        control.serverSocket.close();
        serverThread.join();
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 7]
```diff
@@ -0,0 +1,427 @@
...
+    }
+
+    @Test
+    public void testNonChunkedRequestAndNoExpect100ContinueResponse() throws Exception {
+        String body = "testNonChunkedRequestAndNoExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:182]:
```suggestion
        String body = "testNonChunkedRequestAndNoExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testNonChunkedRequestWithExpect100ContinueResponse() throws Exception {
+        String body = "testNonChunkedRequestWithExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:207]:
```suggestion
        String body = "testNonChunkedRequestWithExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testNonChunkedRequestWithDoubleExpect100ContinueResponse() throws Exception {
+        String body = "testNonChunkedRequestWithDoubleExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:232]:
```suggestion
        String body = "testNonChunkedRequestWithDoubleExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 10]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testChunkedRequestAndNoExpect100ContinueResponse() throws Exception {
+        String body = "testChunkedRequestAndNoExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:257]:
```suggestion
        String body = "testChunkedRequestAndNoExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 11]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testChunkedRequestWithExpect100ContinueResponse() throws Exception {
+        String body = "testChunkedRequestWithExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:283]:
```suggestion
        String body = "testChunkedRequestWithExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 12]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testChunkedRequestWithDoubleExpect100ContinueResponse() throws Exception {
+        String body = "testChunkedRequestWithDoubleExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:309]:
```suggestion
        String body = "testChunkedRequestWithDoubleExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testFixedLengthRequestAndNoExpect100ContinueResponse() throws Exception {
+        String body = "testFixedLengthRequestAndNoExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:335]:
```suggestion
        String body = "testFixedLengthRequestAndNoExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testFixedLengthRequestWithExpect100ContinueResponse() throws Exception {
+        String body = "testFixedLengthRequestWithExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:361]:
```suggestion
        String body = "testFixedLengthRequestWithExpect100ContinueResponse";
        Control control = this.control;
```

File: test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java [hunk 15]
```diff
@@ -0,0 +1,427 @@
...
+        System.err.println("response body: " + responseBody);
+        assertTrue(responseCode == 200,
+                String.format("Expected 200 response, instead received %s", responseCode));
+        assertTrue(body.equals(responseBody),
+                String.format("Expected response %s, instead received %s", body, responseBody));
+    }
+
+    @Test
+    public void testFixedLengthRequestWithDoubleExpect100ContinueResponse() throws Exception {
+        String body = "testFixedLengthRequestWithDoubleExpect100ContinueResponse";
```
Comment (dfuch) [inline test/jdk/java/net/HttpURLConnection/HttpURLConnectionExpectContinueTest.java:387]:
```suggestion
        String body = "testFixedLengthRequestWithDoubleExpect100ContinueResponse";
        Control control = this.control;
```

**Review by dfuch — COMMENTED** [review]
Since control can be accessed from different threads it's better to make it volatile too - and then stuck it in a local variable before reading its fields.

**Review by dfuch — APPROVED** [review]
LGTM

**Review by djelinski — APPROVED** [review]

**Pull request closed without merging**